airsonic-main/src/main/java/org/airsonic/player/controller/SonosSettingsController.java

```diff
@@ -50,46 +51,42 @@
     private SonosService sonosService;
 
     @RequestMapping(method = RequestMethod.GET)
-    public String doGet(Model model) throws Exception {
-
-        Map<String, Object> map = new HashMap<String, Object>();
-
-        map.put("sonosEnabled", settingsService.isSonosEnabled());
-        map.put("sonosServiceName", settingsService.getSonosServiceName());
-
-        model.addAttribute("model", map);
-        return "sonosSettings";
+    public ModelAndView doGet(HttpServletRequest request, Model model) throws Exception {
+        return new ModelAndView("sonosSettings", "model", getModel(request));
     }
 
     @RequestMapping(method = RequestMethod.POST)
-    public String doPost(HttpServletRequest request, RedirectAttributes redirectAttributes) throws Exception {
-        handleParameters(request);
-
-        redirectAttributes.addFlashAttribute("settings_toast", true);
-
-        return "redirect:sonosSettings.view";
-    }
+    public ModelAndView doPost(HttpServletRequest request) throws Exception {
 
-    private void handleParameters(HttpServletRequest request) {
         boolean sonosEnabled = ServletRequestUtils.getBooleanParameter(request, "sonosEnabled", false);
         String sonosServiceName = StringUtils.trimToNull(request.getParameter("sonosServiceName"));
         if (sonosServiceName == null) {
             sonosServiceName = "Airsonic";
         }
 
+        settingsService.setSonosLinkMethod(request.getParameter("sonosLinkMethod"));
         settingsService.setSonosEnabled(sonosEnabled);
         settingsService.setSonosServiceName(sonosServiceName);
+        settingsService.setSonosCallbackHostAddress(request.getParameter("callbackHostAddress"));
         settingsService.save();
 
-        sonosService.setMusicServiceEnabled(false, NetworkService.getBaseUrl(request));
-        sonosService.setMusicServiceEnabled(sonosEnabled, NetworkService.getBaseUrl(request));
-    }
+        List<String> returnCodes = sonosService.setMusicServiceEnabled(sonosEnabled, settingsService.getSonosCallbackHostAddress());
+
+
+        Map<String, Object> map = getModel(request);
+
+        map.put("returnCodes", returnCodes);
 
-    public void setSettingsService(SettingsService settingsService) {
-        this.settingsService = settingsService;
+        return new ModelAndView("sonosSettings", "model", map);
     }
 
-    public void setSonosService(SonosService sonosService) {
-        this.sonosService = sonosService;
+    private  Map<String, Object> getModel(HttpServletRequest request){
+        Map<String, Object> map = new HashMap<String, Object>();
```

Comment:
No need to specify the type in the HashMap, since it'll be automatically inferred.